### ISSUE TITLE
Embed SVG images using <object> instead of <img>

### ISSUE THREAD

**Issue by pepijnve** [asciidoctor/asciidoctor#1301]
The HTML5 backend always generates `<img>` tags for image nodes (see https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/converter/html5.rb#L523). For SVG images it would be nice to have the option of generating `<object>` tags instead to allow hyperlinks inside the SVG images to work. This would resolve https://github.com/asciidoctor/asciidoctor-diagram/issues/64.

**Comment by mojavelinux** [comment]
This strikes me as very odd that the browser forces us to use a different tag in order for the SVG to work properly. Is this defined in the HTML5 spec this way or is this just a browser limitation / bug?

If we move to `<object>` for SVG, I'd prefer to do it full sail. I don't really like the idea of exposing that at a higher level. The output should "just work". If `<img>` is a bad way to embed an SVG, let's move away from it.

**Comment by mojavelinux** [comment]
It appears we need to use the following:

```
<object type="image/svg+xml" data="diagram.svg"/>
```

I suppose we could make it so that an extension can instruct the image block as to which element to use for SVG. That's one possibility. The question is whether there's any advantage at all to using the `<img>` tag. I suppose we get some styles for free that way.

We'll probably want to add `page-break-inside:avoid` and `max-width:100%!important` to the print styles for object. Maybe the `vertical-align:middle` and `display:inline-block` as well for normal styles.

**Comment by mojavelinux** [comment]
Interesting to note that a closing `</object>` tag *must* be used or it cuts off all subsequent content in Chrome.

**Comment by mojavelinux** [comment]
Seems like a fairly straightforward change. Those stylesheet changes seem to do it, though we should probably restrict them to the svg data type just so that we don't go styling things we shouldn't:

for screen:

```
object[type="image/svg+xml"] {
  vertical-align:middle;
  display:inline-block;
}
```

for print:

```
object[type="image/svg+xml"] {
  page-break-inside: avoid;
  max-width:100%!important;
}
```

One thing I noticed is that you can't wrap an `<a>` around an object, so if the image block has a link (a linkable image, not a link inside the image), then we have to fall back to the `<svg>` tag.

**Comment by mojavelinux** [comment]
We also need to decide how we'll know the target is an SVG. I think it's fair to say that we'll base it on the file extension. Sometimes, URLs don't use file extension, so maybe we need an explicit `type` attribute to use as a fallback to handle those cases.

The issue of image type raises the need for an API method to access the image type so the client doesn't have to parse the target every time. When we do that, we should probably also add a method for target as requested in #1357.

**Comment by mojavelinux** [comment]
> for screen:
>
> ```
> object[type="image/svg+xml"] {
>   vertical-align:middle;
>   display:inline-block;
> }

I suggest these styles because they match the styles that are applied to the `img` element in the default Asciidoctor stylesheet. See https://github.com/asciidoctor/asciidoctor/blob/2893d1666060fbb360e717fd503e06f8cc0f6f32/data/stylesheets/asciidoctor-default.css#L59

**Comment by mojavelinux** [comment]
In order to allow a link to be added around the SVG, we need additional styles:

```
.imageblock a.image {
  display: inline-block;
}
a.image object[type="image/svg+xml"] {
  pointer-events: none;
}
```

Of course, this will disable pointer events in the SVG, which is the whole idea of using the `<object>` element. Perhaps in this case we should switch back to using an `<img>` element.

**Comment by mojavelinux** [comment]
We probably don't need the `type="image/svg+xml"` for the print styles and for the link workaround.

**Comment by mojavelinux** [comment]
We'll need to make sure this doesn't break SVGs in the html-pipeline (used on GitHub).

**Comment by pepijnve** [comment]
@mojavelinux perhaps it should be possible to flag an image block as interactive (or something like that) via an attribute and only use object in that case. Opt-in rather than opt-out in other words.

**Comment by mojavelinux** [comment]
That thought crossed my mind while I was implementing it. We could do something like:

```
[%interactive]
image::diagram.svg[]
```

and

```
image:diagram.svg[opts=interactive]
```

Do you like this better or would you prefer that it is sorted out automatically?

NOTE: We'd still need the GitHub hack since this always has to be disabled for SVGs to work on GitHub.

**Comment by mojavelinux** [comment]
Btw, a PR is now available for review.

**Comment by mojavelinux** [comment]
Another thing to consider. While testing, I found that the SVG actually looks better when embedded with an `<object>` element instead of an `<img>` element because it somehow selects the correct fonts in more cases. So this is more than just about interactivity.

My feeling is that we should move forward with using `<object>` when possible because it's a better experience...despite some hacks we need (the web is built on hacks).

**Comment by pepijnve** [comment]
No strong preference either way myself; these kinds of things are always tough judgement calls in my experience.
Opt-out results in a behaviour change by default for all html output which may cause problems for users. Postprocessing of the html output may no longer pick up svg elements for instance if code is looking for img tags specifically.
Opt-in seems to give better default behaviour (better rendering, interactivity, etc) though.

**Comment by mojavelinux** [comment]
>  these kinds of things are always tough judgement calls in my experience.

I've lost many a night's sleep :)

**Comment by mojavelinux** [comment]
We can always start with opt-in, then make it the default behavior in the future (as it becomes used widely). Then, you can just gradually stop using the explicit option. If people really want it, we can just change that requirement sooner.

I'll update the PR.

**Comment by mojavelinux** [comment]
PR updated! Let me know what you think.

**Comment by pepijnve** [comment]
Looks good. Just needs some additional nil checks for the alt text I think (added comment on PR).

**Comment by stephenhay** [comment]
@mojavelinux: to address the question posted on Twitter, I think it depends on whether you want asciidoctor to be more limiting than what you could do in plain HTML. In plain HTML, you can embed SVGs in a document:

* as an `img`. IMHO this should remain possible, as it is simplest use case. No interaction, obviously.
* as an `object`. Useful mainly due the fact that you can place a fallback between the opening and closing tag, *which should then be possible in asciidoc as well*.
* `embed` and via an `iframe`: I see no advantage to including those in asciidoctor.
* as `<svg>`. Useful as well, see below.

(I left out the possibility of using background images, because that's already possible via custom style sheet.)

In the first two, the images are cached. With `<svg>`, it's not, but there's no extra HTTP request. With `<object>`, your interactions must be in the SVG itself. With `<svg>`, you can script your interactions externally. The structure for using `<svg>` would be e.g.:

    <svg version="1.1" xmlns="http://www.w3.org/2000/svg" ..>     
         
    </svg>

So, for the most flexibility, I think my suggestion would be to:

1. Continue to allow SVGs via `<img>`.
2. *Also* allow SVGs via your revised `<object>` method, *but a fallback should be possible to insert  IMHO*. (Again, noting that you're calling it "interactive" but that's only the case if interactivity is baked in the SVG)
3. *Also* allow SVGs via `<svg>`.

The last two should be *possible* but not thrown in the user's face: got an SVG? Just put it in there like any other image; done. Oh, you want a fallback, or have interactivity in your SVG? `object`. Oh, you want to do external scripting CSS animations? `<svg>`.

Agree that alternatives should be opt-in.

**Comment by mojavelinux** [comment]
Thanks for this very thorough feedback, @stephenhay! This is super helpful input.

I'll start out by addressing the option name that enables the switch from `<img>` to `<object>`. You are correct that the SVG isn't always interactive when attached to an `<object>` element as it depends on what is in the SVG, or if behavior is applied to it in some other way. What we want to imply is that it _can_ be interactive. In that case, perhaps a better name for the option is simply "active".

```
image:diagram.svg[opts=active]
```

I think that's a little less suggestive of it actually being interactive and says that it _may_ be more active than the static `<img>` element allows it to be. (I'd also go with `dynamic`). What do you think? If you have another suggestion, we'll definitely consider it.

I really like the idea of being able to set a fallback image! Very nice suggestion. We have a precedent of doing something similar with the video tag. If you set the poster attribute, it will use that image if the video cannot be loaded. So we can follow suit here. And fallback seems like the most suitable name in this case. Note that if a fallback is not set, we're currently putting the alt text as the fallback content.

We are definitely thinking about when to embed the content of the SVG directly inside an `<svg>` element. One thing we could do is embed it this way if the `data-uri` attribute is set on the document and the "active" option is set. For instance, the `<svg>` element would be used under these conditions.

```
:data-uri:

image::diagram.svg[opts=active]
```

We may still want an option that isn't tied to the `data-uri` attribute as that affects how other images are handled as well. Perhaps we can simply allow another option:

```
image::diagram.svg[opts=embed]
```

In this way, the options are as follows:

* opts is not set, use `<img>` element
* opts is "active", use `<object>` element
* opts is "embed", use `<svg>` element (but only if safe mode allows reading from disk)

How do you like that suggestion? Should we move forward with it?

One open question that remains is whether the `<object>` element or the `<svg>` element should be used when `data-uri` is set and the "active" option is set. Do we put Base64 encoded data into the `data` attribute or just inline the SVG content inside the `<svg>` element? In other words:

* `data-uri` is set, opts is "active", what to use?

**Comment by stephenhay** [comment]
Oh, man, naming is hard. Just my opinion, but:

* "active" seems even less obvious to me than "interactive"; given the choice of those two, I'd personally opt for "interactive".
* I think "inline" perhaps makes more sense to me than "embed" as an opts attribute. "Embed" may confuse, since it's possible to use `<embed>` to include an SVG in normal HTML.
* Wondering if simply using "object" and "inline" as attrs would suffice, instead of "interactive|active" and "embed". Neither is intuitive, so people would understand usage based on the docs anyway. Using "object" and "inline" closely mirrors the HTML equivalents, but we still have the benefit of the concise asciidoctor syntax.

I would advocate keeping `:data-uri:` independent of the decision regarding inline or object SVGs. So if I opt for inline, I want `<svg>`, *not* Base64 encoded.

Even with `:data-uri:`, I see no need whatsoever to Base64 inline SVG. It's already XML and part of the document; is there an advantage to doing so? See also: https://css-tricks.com/probably-dont-base64-svg/

Again, just opinions here; I'm no expert on this stuff.

**Comment by mojavelinux** [comment]
> I think "inline" perhaps makes more sense to me than "embed" as an opts attribute.

I love it. Let's do it.

**Comment by mojavelinux** [comment]
> I would advocate keeping :data-uri: independent of the decision regarding inline or object SVGs. So if I opt for inline, I want <svg>, not Base64 encoded.

:+1: Exactly the decisiveness I was looking for!

**Comment by mojavelinux** [comment]
> Even with :data-uri:, I see no need whatsoever to Base64 inline SVG.

The only real benefit (and the whole point of the data-uri feature in AsciiDoc) is to consolidate everything into a single HTML document. It's mainly used for being able to e-mail around an HTML file or something of the sort (and I've certainly used it for that end). Outside that use case, though, I don't really see data-uri as a good permanent publishing solution anyway...so I don't think we should fret about whether we should or shouldn't Base64 encoded the `<svg>`.

> Using "object" and "inline" closely mirrors the HTML equivalents, but we still have the benefit of the concise asciidoctor syntax.

This seems to be the winner, especially since the sole purpose of this feature is to get low-level control of the HTML output.

Since they are mutually exclusive, I wonder if an attribute is better than an option. An option tends to be a switch. So something like:

```
image::diagram.svg[html=object]

image::diagram.svg[html=inline]
```

Other names for this attribute might be:

* html-element
* inlay
* insert-as
* as

Thoughts? Or do we like this better?

* opts=inline
* opts=object

**Comment by stephenhay** [comment]
Several nice name choices there. I kind of like "as". On the other hand, if there are more options, then opts might be simpler, e.g. `image::diagram.svg[opts=inline, foo, bar, this, that]` as opposed to having several attributes. Tough one. What's most consistent with the rest of asciidoc? Consistency will keep usability high. Perhaps "opts" after all?

**Comment by pepijnve** [comment]
I like the self explanatory-ness of `image::diagram.svg[insert-as=object]`.

`object` is HTML specific nomenclature though. I would prefer something like `interactive` over `object`. One is declarative, describing intent. The other describes the means by which the intent is achieved which might change over time.

**Comment by mojavelinux** [comment]
I let my mind sleep on it (yes, sleep) and came up with another term that is perhaps more fitting.

* operative - functioning; having effect

I think this may be spot on to what putting an SVG into an `<object>` does to it. It makes it functional, instead of static. It also allows it to be found in the DOM, so it affects the DOM.

I don't think we have to be overly concerned with the fact that the two options are mutually exclusive as there are other cases of mutually exclusive options controls by the options attribute. Thus, here's the latest proposal.

Link as `<img>`

```
image::diagram.svg[]
```

Link as `<object>`

```
image::diagram.svg[opts=operative]
// or
image::diagram.svg[opts=operative, fallback=diagram.png]
```

Embed as raw `<svg>`

```
image::diagram.svg[opts=inline]
```

What does the jury say?

**Comment by stephenhay** [comment]
I suppose Id be okay with operative, though it feels a bit like going through hoops to avoid HTML-specific naming conventions (even though these options exist a.t.m. purely for HTML). I do tend to agree with @pepijnve though: where possible it's best to describe meaning/intent. In that case, operative seems okay.

The only thing that still kind of bothers me is the fact that a "raw" SVG can also be operative in the same ways as `<object>`-embedded SVGs.

tl;dr: I'm stumped, and as long as no one else has a better idea, your current best idea is the best. :-)

[1] (`[opts=raw]` might be a possibility as opposed to `[opts=inline]` BTW)

**Comment by mojavelinux** [comment]
Just for some background information, it's interesting to note how reStructuredText handles SVG. http://docutils.sourceforge.net/test/functional/expected/standalone_rst_html4css1.html#svg-images

**Comment by mojavelinux** [comment]
It's interesting to note that when we open up this level control, we'll likely be offering the best integration with SVG of any lightweight markup / documentation language (that I know of).

**Comment by mojavelinux** [comment]
I realized that instead of checking for `env-github` to know when using `<img>` instead of `<object>` or `<svg>` (the "secure" scenario), we should base the decision off the safe mode. When safe mode is `:secure` or stricter, we should use `<img>` regardless of the hints from the AsciiDoc syntax.

**Comment by mojavelinux** [comment]
Another alternative for "raw" is "native".

**Comment by mojavelinux** [comment]
A close synonym of "operative" is "functional", which may be more familiar to and comfortable for the technical audience.

**Comment by mojavelinux** [comment]
The idea of using the term "functional" is that it suggest that it's "just functional enough" to be functional, but still not a "native" component (e.g., `<svg>`).

So we'd have:

* default: img
* functional: object
* native (or raw): svg

**Comment by stephenhay** [comment]
I'm no expert, but it sounds good to me.

**Comment by mojavelinux** [comment]
Some relevant discussion about SVG, good for terminology: https://css-tricks.com/using-svg/

**Comment by mojavelinux** [comment]
Some interesting difference between the behavior of an SVG when linked via an object tag vs when it's inline in the HTML: http://benfrain.com/selecting-svg-inside-tags-with-javascript/

**Comment by graphitefriction** [comment]
So in an effort to evaluate our naming choices, I wrote out the benefits and limitations of the two new features (which would output either an `<object>` tag or `<svg>` tag). 

_AKA Documentation Driven Naming_

I didn't really worry about `<img>` since that is our default. It's just here for completeness.

Link as `<img>`

```
image::diagram.svg[]
```
- Can't use CSS interactions or animations (If I remember correctly, it should be able to do CSS animations inside the `<svg>` tag but Asciidoctor nukes this ability (yes?, no?, stop spreading horrible rumors :speak_no_evil: !)

Link as `<object>`

```
image::diagram.svg[opts=operative]
// or
image::diagram.svg[opts=operative, fallback=diagram.png]
```

- SVG DOM is available (Can be accessed by JavaScript, but not JQuery without the addition of other dependencies)
- Can cache
- Control individual SVG elements using CSS, but only from inside the `<svg>` tag
- Can use CSS interactions and CSS animations, but only from inside the `<svg>` tag
- Can add fallback
- Can't control SVG elements or use CSS interactions and animations via an external stylesheet or `<style>` tag from elsewhere in the document

Embed as `<svg>`

```
image::diagram.svg[opts=inline]
```

- Placed in HTML DOM (can be accessed by JQuery)
- No additional HTTP request
- Control individual SVG elements using CSS
- Can use CSS interactions and CSS animations
- CSS can be linked via an external stylesheet or directly in the page via `<style>`

What I noticed from listing the abilities associated with the `<object>` and `<svg>` is that I used the word _control_ a lot.  So I'm throwing the following option names into the pot (or hoops) :grinning: Jump, baby, jump!

Link as `<object>`

```
image::diagram.svg[opts=control]
```

Embed as `<svg>`

```
image::diagram.svg[opts=control-inline]
```

The other word that comes to my mind is that I can _interface_ with these SVGs when using `<object>` and `<svg>`.

Link as `<object>`

```
image::diagram.svg[opts=interface]
```

Embed as `<svg>`

```
image::diagram.svg[opts=interface-inline]
```

**Comment by graphitefriction** [comment]
BTW, found this article about SVGs that shows some great examples of using SVGs in HTML as well as listing the pros and cons of each embedding/linking method: 

http://www.smashingmagazine.com/2014/11/styling-and-animating-svgs-with-css/

**Comment by stephenhay** [comment]
Also, Sara Souiedan has some slides about embedding: http://slides.com/sarasoueidan/working-with-svg-a-primer#/19

> Documentation Driven Naming

:-)

**Comment by mojavelinux** [comment]
Sara is doing a fantastic job educating the world about SVG. These are amazing resources to have for the Asciidoctor User Manual.

Asciidoctor :heart: SVG!

**Comment by mojavelinux** [comment]
We're nearing the release of 1.5.3 and this is one of the remaining issues, so it's decision time! I've come back into this with fresh eyes and reviewed the terminology in the linked articles. Here's what I think makes the most sense.

Clearly `inline` is the familiar way to describe putting the SVG data directly in the HTML document inside an `<svg>` element, so that seems to be a clear winner. Sara S. uses that terminology consistently as well.

The word `operative` just looks funny now that I come back to it. My gut tells me to choose "functional" instead. In a way, I think it maps well to `<object>` because in programming we think of an object as having functions.

I thought about the idea of "control", but since it's not an adjective it's not parallel (it would need to be "controllable", which perhaps is too wordy).

In the future, we may even decide to use the `<object>` by default if the safe mode allows it. That would reduce the choice down to just `inline`, so we can keep that in mind.

* img (default)
* object (opts=functional)
* svg (opts=inline)

I'm going to go ahead and update the pull request to implement this proposal. In the meantime, I'll give you a chance to speak now or, well, you know the rest :)

**Comment by pepijnve** [comment]
Based on the options that are left and a quick dictionary lookup I would go for `interactive` over `functional`.

From the Apple Dictionary app:
- Interactive:
  1. allowing a two-way flow of information between a computer and a computer-user; responding to a user’s input: a fully interactive map of the area.
- Functional:
  1. of or having a special activity, purpose, or task: a functional role.
  2. designed to be practical and useful, rather than attractive: a small, functional bathroom.
  3. in operation; working: the museum will be fully functional from the opening of the festival.

The third meaning for functional is what I think you mean. That would be ok for me if the alternative was completely static, but it's not. SVG animations will still work if you use an `img` tag. User interaction with the SVG will not though.

**Comment by mojavelinux** [comment]
I'm inclined to agree. It was our original instinct after all. I've implemented inclusion using the `<object>` when the option value is "interactive". The PR is now updated.

I also implemented the "inline" behavior as part of the fix for this issue. I encountered a limitation however. Right now, it can only read the SVG data from a local file. In other words, we don't yet have support for embedding the SVG data from a remote URI. For example, the following won't work:

```
image::https://travis-ci.org/asciidoctor/asciidoctor.svg[Build Status,opts=inline]
```

Only the alt text will be shown.

I need to make new methods that handle that logic because otherwise it's surprisingly complex. I might do that in a follow-up issue since it probably isn't critical (you can work around it).

**Comment by mojavelinux** [comment]
I've implemented it so that the SVG data is inserted inline when `opts=inline` regardless of the `data-uri` setting. All other permutations will base64 encode the SVG data when the `data-uri` attribute is set (assuming the safe mode allows it).

**Comment by mojavelinux** [comment]
Okay, I went ahead and implemented the retrieval of SVG data from a remote URI :)

I put the code to read SVG data from a URI directly inside the converter. We'll want to migrate it to somewhere later so it can be used by other converters...but it for now, it works.

**Comment by mojavelinux** [comment]
Here's another great resource on SVG fallbacks, btw.

https://css-tricks.com/a-complete-guide-to-svg-fallbacks/

**Comment by pepijnve** [comment]
Great stuff. This is going to be a very useful feature!

**Comment by mojavelinux** [comment]
:+1:

**Comment by mojavelinux** [comment]
I have one last thought before we close the door on this one. I wonder if we should build on this idea of "interactions" by using it as the key instead of the value.

* interactions=basic uses object
* interactions=inline uses svg

That way, we have degrees of interactivity, which is really what is happening when we change from one method of embedding to the next. It also suggests the level of intermingling (interactions) between the two document models.

**Comment by pepijnve** [comment]
I think that would conflate two intentions that should remain distinct. You want interactiveness if you're SVG is an interactive one. You want to inline if the resulting document should be self-contained. One might end up implying the other, but I'm not sure it's a good idea to tie them together. You could for instance implement things as follows:
- inline -> img with data uri
- inline, interactive -> svg or object with data uri
- interactive -> object with non-data uri
- nothing -> img with non-data uri

The remaining bit that isn't covered by this is if the SVG should be styleable by the CSS of the containing document. This requires embedding with an svg tag. No good suggestion for how to express that yet. Perhaps `styleable`?

**Comment by mojavelinux** [comment]
> Orthogonal to that there's the option of wanting a standalone document or not which is what inline is for.

That's not quite how I see it. When the goal is to simply create a standalone document (or mostly standalone atm), I think the `data-uri` is sufficient. You don't really care how it's done, just that it's all one document. When the “how” matters, because it changes the behavior or capabilities, then you need to be more specific and use the hints. The inline hint brings the SVG data into the document (`<svg>` element), the interactive hint merely adds what can be added of interactivity while still linking to the resource (`<object>` element).

In summary, I do believe that we are talking about levels of behavior.

> it makes the most sense to tag the image in asciidoc with the set of capabilities you want as directly as possible.

I agree with this in theory. In practice, though, we don't want to annoy the author by making it unnecessarily verbose. I think a good API is as specific as necessary for the human to “get it”. Any more and it's just pageantry.

A possible compromise is to require both options to get inline behavior:

```
image::sample.svg[opts="interactive,inline"]
```

also written as:

```
[%interactive%inline]
image::sample.svg[]
```

We could recognize "inline" as a shorthand (for those who really want to type less) but typing it out longhand would be permitted for clarity. (Building on something Sarah had suggested).

This seems like a reasonable compromise (for something that is turning out to be extremely difficult to define because there aren't well-defined lines or existing terminology). This way, we understand that "inline" represents all features that are enabled when the SVG data is imported into the document. We know the `<object>` element only gives us interactivity. (If there's a word other than "interactive" that can explain what `<object>` give it, I'm open to it).

**Comment by mojavelinux** [comment]
On a side note, should the attribute for the fallback image be `fallback-image` instead of `fallback` (since it's not clear what type of fallback should be specified)? As it's currently implemented, the alt text is used as fallback text if a fallback image is not specified.

**Comment by mojavelinux** [comment]
To clarify my point about using both inline and interactive, if "inline" is present, it wins. Then it looks for "interactive". That remains consistent with the semantics.

**Comment by mojavelinux** [comment]
I'm pretty happy with the PR at this point. Feel free to review.

I even managed to extract the method that reads the SVG data (the generic part) so it can be used in other converters. I think we've got ourselves a solid starting point.

**Comment by mojavelinux** [comment]
After obsessively checking all the corner cases and performance of this pull request and analyzing the implementation to death, I'm thrilled to say that I've finally pulled the trigger and merged these changes!!! That also means we're ready for the 1.5.3 release.

Thanks @pepijnve, @stephenhay and @graphitefriction for providing valuable input and insight on this critical issue (and Sara S. for all the amazing SVG articles). This wasn't an easy issue to solve by any stretch, but the good ones never are. :beers: :wine_glass: 

I'll be anxiously awaiting to see how people put this to use and what else we can learn along the way.

**Comment by pepijnve** [comment]
Time for a mini celebration 🎉. Good stuff.

**Comment by mojavelinux** [comment]
Trust me, I've got the St Bernardus Abt 12 queued up. It's been in the fridge since the summer. Breaking it out Friday night for some release action.

**Comment by pepijnve** [comment]
You can get St Bernardus in the states? Color me surprised. Or is it one of the benefits of visiting Devoxx Belgium? :)

**Comment by mojavelinux** [comment]
Yep, the import like crazy over here. Granted, it's not as good as having it at Bier Central, but that's why there's Devoxx.

**Comment by astefanutti** [comment]
I've tried inlining SVG output by PlantUML Asciidoctor Diagram extension within a DZSlides backend based presentation without success. Ideally, I'd like to write something like: 
```
[plantuml, "type-meta", "svg", height="95%", width="100%", opts="inline"]
----
@startuml  
//...
@enduml
----
```

And trying `image::plantuml.svg[opts=inline]` as a work-around is not inlining the SVG neither.

Are the new SVG rendering options available in all HTML based backends?

PS: I find the quality of the discussion and the collaboration in this thread impressive :100:

**Comment by mojavelinux** [comment]
>  I find the quality of the discussion and the collaboration in this thread impressive

Thanks @astefanutti!

> Are the new SVG rendering options available in all HTML based backends?

Yes, this is one of those features that requires coordination with the converter. Fortunately, it isn't too much extra work.

Here's what needs to happen.

The line https://github.com/asciidoctor/asciidoctor-backends/blob/master/slim/dzslides/block_image.html.slim#L3 needs to get augmented with the logic from the line range https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/converter/html5.rb#L543-L551.

You'll also need to add the helper from the line range https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/converter/html5.rb#L1144-L1158 to the helpers.rb file.

Can you open an issue in asciidoctor-backends so we can track this change?

**Comment by astefanutti** [comment]
@mojavelinux, thanks for the prompt feedback and the implementation details.

I've just created asciidoctor/asciidoctor-backends#122.

**Comment by mojavelinux** [comment]
:+1:

**Comment by nicorikken** [comment]
This feature is golden ❤️ , allowing to ship a single HTML document with graphics, rather than having to bundle all the images as well. PDF is limiting the content, and MHTML is not yet widely supported.
I tried the example by @astefanutti even though it was a proposal/idea. It seems the `width` needs to be set, even though it is ignored. LIkewise the `"type-meta"` feature doesn't seem to work, but that might be part of the proposal.
If I'm not mistaken this should be an example for a proper attribute: `[plantuml, architecture-diagram, svg, opts="inline", height="100%", width="100%"]`
Now if only it was easy to embed Javascript and/or CSS I would be able to generate full-featured Remark.JS presentations within a single HTML-file. 😄
